velodrome/config.py

```diff
@@ -0,0 +1,103 @@
+# limitations under the License.
+
+#!/usr/bin/env python
+
+import os
+import subprocess
+import sys
+import yaml
+import StringIO
+
+CONFIG = "config.yaml"
+
+DEPLOYMENTS = {
+    "fetcher": "fetcher/deployment.yaml",
+    "transform": "transform/deployment.yaml",
+    "influxdb": "grafana-stack/influxdb.yaml",
+    "prometheus": "grafana-stack/prometheus.yaml",
+    "prometheus-config": "grafana-stack/prometheus-config.yaml",
+    "grafana": "grafana-stack/grafana.yaml",
+    "grafana-config": "grafana-stack/grafana-config.yaml",
+    "nginx": "grafana-stack/nginx.yaml",
+    "sqlproxy": "mysql/sqlproxy.yaml",
+}
+
+
+def main():
+    if len(sys.argv) != 1:
+        print >> sys.stderr, "Too many arguments. Specify"
+        sys.exit(128)
+
+    with open(get_absolute_path(CONFIG)) as config_file:
+        config = yaml.load(config_file)
+        print_deployments(["sqlproxy"], {})
+        for project_name, project in config['projects'].items():
+            public_ip = project.get('nginx', {}).get('public-ip', '') or ''
+            print_deployments(["influxdb", "grafana", "nginx"], {
+                "PROJECT": project_name,
+                "NGINX_PUBLIC_IP": public_ip,
+            })
+            if 'prometheus' in project:
+                print_deployments(["prometheus"], {
+                    "PROJECT": project_name,
+                })
+                patch_configuration("prometheus-config",
+                                    project['prometheus'],
+                                    { "PROJECT": project_name })
+            if 'grafana' in project:
+                patch_configuration("grafana-config",
+                                    project['grafana'],
+                                    { "PROJECT": project_name })
+            for repository in project['repositories']:
+                print_deployments(["fetcher", "transform"], {
+                    "GH_ORGANIZATION": repository.split("/")[0],
+                    "GH_REPOSITORY": repository.split("/")[1],
+                    "PROJECT": project_name,
+                })
+
+
+def patch_configuration(component, values, env):
+    with open(get_absolute_path(DEPLOYMENTS[component])) as f:
+        config = yaml.load(f)
+        # We want to fail if we have unknown keys in values
+        unknown_keys = set(values) - set(config['data'])
+        if unknown_keys:
+            raise ValueError("Unknown keys in config:", unknown_keys)
+        config['data'] = values
+    full_conf = StringIO.StringIO(yaml.dump(config, default_flow_style=False))
+    print_deployment(full_conf, env)
+
+
+def get_absolute_path(path):
+    return os.path.join(os.path.dirname(__file__), path)
+
+
+def print_deployments(components, env):
+    for component in components:
+        with open(get_absolute_path(DEPLOYMENTS[component])) as f:
+            print_deployment(f, env)
+
+
+def print_deployment(f, env):
+    p = subprocess.Popen(["envsubst", " ".join(["$" + key for key in env])],
```

Comment:
I didn't know about this program. Neat. I guess python also has `os.path.expandvars`, but this probably works too.